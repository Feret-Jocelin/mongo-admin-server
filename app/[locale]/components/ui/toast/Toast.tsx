'use client';

import { useEffect, useMemo, useState } from 'react';

import styles from './toast.module.scss';

import SvgIcon from '@components/ui/icon/SvgIcon';

import eventEmitter from "@/shared/emitter/events";

interface ToastProps {
  type: 'success' | 'error';
  message: string;
}

const Toast = () => {
  const [visible, setVisible] = useState(false);
  const [status, setStatus] = useState<ToastProps['type'] | null>(null);
  const [message, setMessage] = useState<ToastProps['message']>('');

  useEffect(() => {
    eventEmitter.subscribe('alert', (event: ToastProps) => {
      setStatus(event.type);
      setMessage(event.message);
      setVisible(true);
      
      setTimeout(() => {
        setVisible(false);
      }, 5000);
    });
  }, [visible]);

  const toastClasses = useMemo(() => {
    const _arr = [styles['toast']];
    if (visible) _arr.push(styles['toast--open']);
    if (status) _arr.push(styles[`toast--${status}`]);

    return _arr.join(' ');
  }, [visible, status]);

  const iconMap = useMemo(() => {
    return {
      success: <SvgIcon className={styles['toast--success--icon']} icon_name="success" />,
      error: <SvgIcon className={styles['toast--error--icon']} icon_name="error" />,
    };
  }, []);

  const toastIcon = status ? iconMap[status] : null;

  return (
    <div className={toastClasses} role="alert">
      <div className={styles['toast--message']}>
        {toastIcon}
        <h3>{message}</h3>
      </div>
      <SvgIcon className={styles['toast--close']} icon_name="close" onClick={() => setVisible(false)} />
    </div>
  );
}

export default Toast;