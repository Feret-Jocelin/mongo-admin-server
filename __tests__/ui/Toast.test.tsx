/* eslint-disable no-undef */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { NextIntlClientProvider } from 'next-intl';
import Toast from '@/app/[locale]/components/ui/toast/Toast';

jest.mock('next-intl/client', () => ({
  usePathname() {
    return '';
  },
  useRouter() {
    return {
      push: () => jest.fn(),
      replace: () => jest.fn(),
    };
  },
}));

const locale = 'fr';
const message = require(`../../shared/locales/${locale}.json`)

describe("FormCreateDB", () => {
  it('renders', () => {
    const handleClose = jest.fn();
    const { container } = render(
      <NextIntlClientProvider locale={locale} messages={message}>
        <Toast
          type='success'
          message='Mock success message'
          onClose={handleClose}
        />
      </NextIntlClientProvider>
     
    );
    expect(container).toMatchSnapshot();
  })
})