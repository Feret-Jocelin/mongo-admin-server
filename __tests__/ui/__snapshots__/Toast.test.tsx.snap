// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCreateDB renders 1`] = `
<div>
  <div
    class="toast toast--open toast--success"
    role="alert"
  >
    <div
      class="toast--message"
    >
      <div
        class="toast--success--icon "
      />
      <h3>
        Mock success message
      </h3>
    </div>
    <div
      class="toast--close pointer-cursor"
    />
  </div>
</div>
`;
